torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1356,6 +1360,33 @@ def _free_full_params_and_use_local_shard(params_to_free):
                         _free_full_params_and_use_local_shard(currently_local_params)
                         self.training_state = TrainingState_.IDLE
 
+    def named_parameters(
+        self,
+        *args,
+        **kwargs,
+    ) -> Iterator[Tuple[str, torch.nn.Parameter]]:
+        """
+        Overrides :meth:`torch.nn.Module.named_parameters()` to intercept
+        parameter names and remove all occurrences of the FSDP-specific
+        flattened parameter prefix when inside the :meth:`summon_full_params`
+        context manager.
+        """
+        # Determine which logic to use based on the context at call time
+        if getattr(self, "training_state", None) != TrainingState_.SUMMON_FULL_PARAMS:
+            for param_name, param in torch.nn.Module.named_parameters(
+                self, *args, **kwargs,
+            ):
+                # Do not modify the behavior if not in `summon_full_params()`
+                yield (param_name, param)
+        else:
+            for param_name, param in torch.nn.Module.named_parameters(
+                self, *args, **kwargs,
+            ):
+                # Remove any instances of the FSDP-specific prefix; there can
+                # be multiple in the case of nested FSDP modules
+                param_name = param_name.replace(FSDP_PREFIX, "")
+                yield (param_name, param)
```

Comment:
Can we do the following to avoid the duplicated for loop?

```
in_summon = (training_state == summon_full_params)
for n, p in named_parameters():
    name = name.replace(...) if in_summon else name
    yield (name, p)
```

Reply:
Great point.